### ISSUE TITLE
Training instability

### ISSUE THREAD

**Issue by DogyunPark** [NVlabs/PixelDiT#6]
Hi,

Thank you for the wonderful work.

I’m currently trying to reproduce the results, but unfortunately, I keep observing unstable training dynamics. For example, after a certain number of iterations, the gradient norm and loss suddenly become abnormally high.

I’m running the PixelDiT-B model on ImageNet 256×256, using the configuration exactly as provided in your GitHub repository.

I was wondering whether this behavior is expected or whether you also observed similar training dynamics during your experiments.

Thank you!

<img width="941" height="509" alt="Image" src="https://github.com/user-attachments/assets/b3555580-a3a4-4cfb-b126-05a27c1d5f23" />

**Comment by frutiemax92** [comment]
The paper suggests to use REPA to stabilize the loss. Personally, I found that clipping the gradients was sufficient to stabilize the training.

**Comment by XZT24** [comment]
you can try switching adam optimizer beta from betas=(0.9, 0.999) to (0.9, 0.95)

**Comment by yeates** [comment]
Hi, thanks. I observe this behavior as well and discuss it in the paper.

I try many approaches to address it and eventually find that the AdaLN modulation scale in the PiT block amplifies values. Switching to post-modulation (see #8 for details) eliminates the issue, as shown in the curve below.

<img alt="image" src="https://github.com/user-attachments/assets/a77ae033-eb24-4b81-bea8-f34ad8eb2c52" />
<img alt="image" src="https://github.com/user-attachments/assets/6aacdd0a-a49f-4fc8-93d9-019c7be4034a" />

I obtain a result of 1.64 from the 320-epoch checkpoint (unswept), which suggests that post-modulation does not significantly affect performance.

Feel free to give it a try. If instability still occurs, please reopen the issue anytime.

**Comment by wxiong-vision** [comment]
More explanation on the motivation and purpose of post-adaln: 

Generally speaking, we find that adaln can amplify feature values significantly. This is affordable in the patch branch as the number of tokens are not too many. However, it becomes a severe issue in the PiT blocks, where each pixel is a token and there are too many tokens. Each feature value is associated with the length of the input tokens, and if the length is large, this may causes some high values in the model. Then adaln will further significantly amplify this value to very large numbers. Then when these values are sent to the attention, the matrix multiplication can even amplify the values and the precision may not afford that, and then gradients explode. Our findings support this hypothesis. We find if we remove the attention layer in PiT blocks, the training will be much more stable. However, the FID drops significantly. 

The purpose of this post-adaln we derive is to resolve this training instability, while preserving the image quality. We find that if we do attention first (in which the values are not significantly amplified before being sent to attention), then do adaln, things are going pretty well and no loss spikes at all. There is still a bit instability at the very beginning, but it is expected and may be further improved by some warmup.  

Stabilizing training is very important in scaling up the whole system and supporting longer iteration. We spent many months to identify and eliminate factors that can cause instability. Let us know if you still find other instabilities, or performance drop, or ways to improve the stability. We would like to discuss with the community more and continue improving the model. 


Wei
